Allow to return connections even if pool was already closed

soemeier opened the pull request:
Fix for #523. I think I have also fixed the problem the you mentioned if the subscriber is being called when the pool is shutDown.

soemeier committed:
Allow to return connections even if pool was already closed

soemeier commented:
Hmm I have checked those tests. They do not even do anything with the connection pool. Might it be that they are unstable?

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -110,54 +120,73 @@ public ConnectionPoolImpl(RxClient.ServerInfo serverInfo, PoolConfig poolConfig,
     public Observable<ObservableConnection<I, O>> acquire() {
 
-        if (isShutdown.get()) {
+        if (isShutdownRequested.get()) {
```
This should be moved inside the `call` method below, so that it gets checked for every subscription of the returned `Observable`.

soemeier replied:
It is already in there but I wanted to fail fast. `if (!lockRetrieved || isShutdownRequested.get())` Do you think it is enough in the call method?

NiteshKant replied:
Yes I think doing it on subscription is enough (and correct) as subscriptions are lazy. So, generally state checks are better on subscription vis-a-vis on creation.

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -54,7 +58,13 @@
+    //is needed to not perform the shutdown twice
+    private final AtomicBoolean isShutdownPerformed = new AtomicBoolean();
+    //is needed to not perform shutdown during the aquire of a connection
+    private final ReadWriteLock shutdownLock = new ReentrantReadWriteLock();
```
So, I don't think we need to be absolutely sure that shutdown does not happen during acquire. There are the following reasons for it:
- There could always be already checked out connections which aren't too different from concurrent acquire and shutdown.
- This lock means that the `shutdown` can not proceed while acquires are happening. If the shutdown happens from the eventloop (it can for rejection of hosts), it blocks the eventloop.

I would suggest having a boolean guarded by `synchronized` which is the flag for shutdown. If `acquire` finds this flag to be `true` before acquire, it is rejected.

`release` should cleanup when `aquiredConnectionsCounter` is 0 and the pool is shutdown.

The above approach will have some connections, when `release` happens after `shutdown` but between `acquire` checking for `shutdown` and actually acquiring, that will be checkout even when the pool is cleaned up. I think we are ok with such behaviors if we check for `shutdown` within `onSubscribe` (which I have suggested below). WDYT?

soemeier replied:
> > There could always be already checked out connections which aren't too different from concurrent acquire and shutdown.

Shouldn't it be that in that case the aquiredConnectionsCounter not equal to zero and the shutdown would not take place then?

> This lock means that the shutdown can not proceed while acquires are happening. If the shutdown happens from the eventloop (it can for rejection of hosts), it blocks the eventloop.

I see your point. Blocking in the event loop is not a good idea.

> release should cleanup when aquiredConnectionsCounter is 0 and the pool is shutdown.

That is what it is doing in my implementation or did I miss something?

> The above approach will have some connections, when release happens after shutdown but between acquire checking for shutdown and actually acquiring, that will be checkout even when the pool is cleaned up. I think we are ok with such behaviors if we check for shutdown within onSubscribe

I somehow do not like the idea that there is still the possibility of a connection leak. I know that this will be a rare case but it can still happen. What about the following approach:
Instead of blocking in the `shutdown` method we just do nothing if we can't acquire the lock. If we can't acquire the lock it basically means that someone is just about to acquire a connection. That connection will at some time be returned and the connection pool is closed.
What could still happen is that acquiring the connection fails. In that case the connection would not be returned and the pool would not be closed. So if the acquire fails we check if we have to shut down the pool. This way there is no chance to return a connection from an already closed connection pool.

NiteshKant replied:
I kind of like your approach if it was without the lock :) I somehow don't like the idea of locking for a non-critical operation. Having said that, it may be clearer to me if I see the code.

So, there isn't a connection leak but a discrepancy in metrics, rite?

soemeier replied:
Well it is not a real leak. So the connections get returned. But ribbon uses a global counter. So all the pools share the same counter in order to have a general connection limit. If that counter does not get decremented due to wrong metrics we get pool exhausted at some point.

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -169,23 +198,27 @@ public void call(final Subscriber<? super ObservableConnection<I, O>> subscriber
                 discardConnection(connection);
                 metricEventsSubject.onEvent(ClientMetricsEvent.POOL_RELEASE_SUCCESS, Clock.onEndMillis(
                         startTimeMillis));
+                performShutdownIfRequested();
```
I think it will be better to call this when `aquiredConnectionsCounter.decrementAndGet();` above returns `0`

soemeier replied:
I will change that.

NiteshKant commented:
@soemeier yes the tests are flaky. Thanks for the changes

soemeier committed:
#525 Use a seperate thread for checking if the pool can be shut down.
Also lock during creation of connection since this is done async in a different thread

soemeier commented:
I did a lot of thinking about it and came to the conclusion that the shutdown on demand just blows up the code. For readability I think it is better to create a separate thread that tries to shutdown the pool on a regular base. Not very nice since we have to create another thread but I think it is more readable. Maybe you have a better approach then using a new thread.

NiteshKant commented:
@soemeier apologies I haven't been able to look at your recent changes. Perhaps, will do so in a couple of days. Thanks for putting this effort to fix ❤️

soemeier committed:
Improvement of the connection pool

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -210,17 +250,64 @@ public void call(final Subscriber<? super ObservableConnection<I, O>> subscriber
         }
+        boolean shutdownPerformed = performShutdownIfPossible();
+        if (!shutdownPerformed) {
+            ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
```
I would rather use `Observable.timer()` instead of creating our own Schedulor (looks like for each shutdown?)

soemeier replied:
I changed this to Observable. Seems to be the better solution. Please have a look at it.

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -110,75 +123,100 @@ public ConnectionPoolImpl(RxClient.ServerInfo serverInfo, PoolConfig poolConfig,
-                    subscriber.onError(throwable);
+                    performAquire(subscriber);
+                }   finally {
+                    lock.unlock();
```
Since connect is asynchronous, this `try-finally` isn't really guaranteeing that there won't be a pending `connect` when `lock` is released.

soemeier replied:
That is true but I increment the connections counter before I release the lock (in the previous commit it was a different counter but that is actually not needed). So if the lock is released the shutdown can not take place because the counter got incremented.

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -110,75 +123,100 @@ public ConnectionPoolImpl(RxClient.ServerInfo serverInfo, PoolConfig poolConfig,
     }
 
+    private void performAquire(Subscriber<? super ObservableConnection<I, O>> subscriber) {
+        logger.trace("Started to aquire connection");
```
I don't think we need these trace logs.

soemeier replied:
I have removed them. They just helped me to understand the problem better ;).

soemeier committed:
Remove traces and use Observables instead of Thread

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -210,17 +243,64 @@ public void call(final Subscriber<? super ObservableConnection<I, O>> subscriber
+        Observable.just(1L).subscribe(createShutdownAction());
+    }
+
+    private void performShutdownIfRequested() {
```
Looks like this method is not used anymore.

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -273,12 +354,22 @@ public void call(ObservableConnection<I, O> connection) {
         );
     }
 
+    private void poolAlreadyClosed(PooledConnection<I, O> connection, long startTime, Subscriber<? super ObservableConnection<I, O>> subscriber) {
```
Method not used anymore?

NiteshKant commented on rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -289,18 +380,47 @@ public Subscription subscribe(MetricEventsListener<? extends ClientMetricsEvent<
         }
     }
+
+    private class ShutdownTask implements Runnable {
```
class not used anymore?

NiteshKant commented:
Apologies again @soemeier for the delayed responses, I am still on PTO. 
I think we can clean up the code a bit (unused classes/methods) then I can pull it in. I am putting this up for the next milestone.

NiteshKant commented:
@soemeier I am pulling this in and will do the cleanup myself. Thanks for the contribution and apologies for the delay due to me being on PTO.
